### PR TITLE
Added support for custom CA bundles in TLS connections to teacher model

### PR DESCRIPTION
**Pull request by relyt0925**

### CODE REVIEW

File: src/instructlab/sdg/generate_data.py [hunk 1]
```diff
@@ -204,6 +204,7 @@ def generate_data(
     tls_client_passwd: Optional[str] = None,
     # TODO need to update the CLI to specify which pipeline to use (simple or full at the moment)
     pipeline: Optional[str] = "simple",
+    tls_ca_bundle: Optional[str] = None,
```
Comment (russellb) [inline src/instructlab/sdg/generate_data.py:207]: I had to look up the httpx.Client API docs to find out for sure what valuer was supposed to be used here (a path to a CA bundle). Could you add a comment above this or in a docstring that clarifies this?  You could link to https://www.python-httpx.org/advanced/ssl/ which is where I found the answer.

**Review by russellb — COMMENTED** [review]
One request for an additional comment, but otherwise lgtm, thanks!

File: src/instructlab/sdg/generate_data.py (hunk 1 above)
Comment (relyt0925) [reply src/instructlab/sdg/generate_data.py:209]: updated with comment!

**Comment by relyt0925** [comment]
(running e2e test off custom build of ilab right now)

**Review by russellb — DISMISSED** [review]
Thanks for the quick update and for the PR!

I also filed this issue to expose this via the CLI -- https://github.com/instructlab/instructlab/issues/1647

**Comment by relyt0925** [comment]
Happy path of CA cert
```
[root@rhel-ai-sdg-training-client-1 instructlabbigdisk]# ilab generate --taxonomy-path taxonomy/compositional_skills/extraction/email/markdown/qna.yaml --endpoint-url https://781d2e7c-us-east.lb.appdomain.cloud/v1 --model-family mixtral --num-instructions 1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1 --tls-ca-bundle teachermodel.crt 
ilab generate --taxonomy-path taxonomy/compositional_skills/extraction/email/markdown/qna.yaml --endpoint-url https://781d2e7c-us-east.lb.appdomain.cloud/v1 --model-family mixtral --num-instructions 1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1 --tls-ca-bundle teachermodel.crt --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1
You are using an aliased command, this will be deprecated in a future release. Please consider using `ilab data generate` instead
INFO 2024-07-10 01:39:40,296 utils.py:145: _init_num_threads Note: detected 80 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" environment variable.
INFO 2024-07-10 01:39:40,296 utils.py:148: _init_num_threads Note: NumExpr detected 80 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 16.
INFO 2024-07-10 01:39:40,296 utils.py:161: _init_num_threads NumExpr defaulting to 16 threads.
INFO 2024-07-10 01:39:40,483 config.py:58: <module> PyTorch version 2.3.1 available.
Generating synthetic data using '/instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1' model, taxonomy:'taxonomy/compositional_skills/extraction/email/markdown/qna.yaml' against https://781d2e7c-us-east.lb.appdomain.cloud/v1 server
INFO 2024-07-10 01:39:41,145 generate_data.py:266: generate_data Synthesizing new instructions. If you aren't satisfied with the generated instructions, interrupt training (Ctrl-C) and try adjusting your YAML files. Adding more examples may help.
INFO 2024-07-10 01:39:41,155 pipeline.py:48: generate Running block: gen_skill_grounded
INFO 2024-07-10 01:39:41,155 pipeline.py:49: generate Dataset({
    features: ['task_description', 'seed_context', 'seed_question', 'seed_response'],
    num_rows: 3
})
INFO 2024-07-10 01:39:48,014 generate_data.py:294: generate_data Generated 1 samples
INFO 2024-07-10 01:39:48,015 generate_data.py:312: generate_data Generation took 6.90s
[root@rhel-ai-sdg-training-client-1 instructlabbigdisk]# 
```

**Comment by relyt0925** [comment]
Showing it fails without specifying the TLS ca (valid as it's signed by private CA)
```
[root@rhel-ai-sdg-training-client-1 instructlabbigdisk]# ilab generate --taxonomy-path taxonomy/compositional_skills/extraction/email/markdown/qna.yaml --endpoint-url https://781d2e7c-us-east.lb.appdomain.cloud/v1 --model-family mixtral --num-instructions 1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1 
ilab generate --taxonomy-path taxonomy/compositional_skills/extraction/email/markdown/qna.yaml --endpoint-url https://781d2e7c-us-east.lb.appdomain.cloud/v1 --model-family mixtral --num-instructions 1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1
You are using an aliased command, this will be deprecated in a future release. Please consider using `ilab data generate` instead
INFO 2024-07-10 01:40:27,293 utils.py:145: _init_num_threads Note: detected 80 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" environment variable.
INFO 2024-07-10 01:40:27,293 utils.py:148: _init_num_threads Note: NumExpr detected 80 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 16.
INFO 2024-07-10 01:40:27,293 utils.py:161: _init_num_threads NumExpr defaulting to 16 threads.
INFO 2024-07-10 01:40:27,479 config.py:58: <module> PyTorch version 2.3.1 available.
Generating synthetic data using '/instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1' model, taxonomy:'taxonomy/compositional_skills/extraction/email/markdown/qna.yaml' against https://781d2e7c-us-east.lb.appdomain.cloud/v1 server
INFO 2024-07-10 01:40:28,169 generate_data.py:266: generate_data Synthesizing new instructions. If you aren't satisfied with the generated instructions, interrupt training (Ctrl-C) and try adjusting your YAML files. Adding more examples may help.
INFO 2024-07-10 01:40:28,179 pipeline.py:48: generate Running block: gen_skill_grounded
INFO 2024-07-10 01:40:28,179 pipeline.py:49: generate Dataset({
    features: ['task_description', 'seed_context', 'seed_question', 'seed_response'],
    num_rows: 3
})
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
    yield
  File "/usr/local/lib/python3.11/site-packages/httpx/_transports/default.py", line 233, in handle_request
    resp = self._pool.handle_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py", line 216, in handle_request
    raise exc from None
  File "/usr/local/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py", line 196, in handle_request
    response = connection.handle_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpcore/_sync/connection.py", line 99, in handle_request
    raise exc
  File "/usr/local/lib/python3.11/site-packages/httpcore/_sync/connection.py", line 76, in handle_request
    stream = self._connect(request)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpcore/_sync/connection.py", line 154, in _connect
    stream = stream.start_tls(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpcore/_backends/sync.py", line 152, in start_tls
    with map_exceptions(exc_map):
  File "/usr/lib64/python3.11/contextlib.py", line 158, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.11/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 978, in _request
    response = self._client.send(
               ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 914, in send
    response = self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 942, in _send_handling_auth
    response = self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 979, in _send_handling_redirects
    response = self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 1015, in _send_single_request
    response = transport.handle_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/httpx/_transports/default.py", line 232, in handle_request
    with map_httpcore_exceptions():
  File "/usr/lib64/python3.11/contextlib.py", line 158, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.11/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/ilab", line 8, in <module>
    sys.exit(ilab())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/utils.py", line 551, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/data/generate.py", line 202, in generate
    generate_data(
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/generate_data.py", line 288, in generate_data
    new_generated_data = sdg.generate(ds)
                         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/sdg.py", line 19, in generate
    dataset = pipeline.generate(dataset)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/pipeline.py", line 51, in generate
    dataset = block.generate(dataset, **gen_kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/iterblock.py", line 23, in generate
    batch_generated = self.block.generate(
                      ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/llmblock.py", line 123, in generate
    outputs = [self._generate([sample], **gen_kwargs)[0] for sample in samples]
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/llmblock.py", line 123, in <listcomp>
    outputs = [self._generate([sample], **gen_kwargs)[0] for sample in samples]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/instructlab/sdg/llmblock.py", line 94, in _generate
    response = self.client.completions.create(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_utils/_utils.py", line 277, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/resources/completions.py", line 528, in create
    return self._post(
           ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1266, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 942, in request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1002, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1079, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1002, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1079, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/openai/_base_client.py", line 1012, in _request
    raise APIConnectionError(request=request) from err
openai.APIConnectionError: Connection error.
```

**Comment by relyt0925** [comment]
tls_insecure still works
```
ilab generate --taxonomy-path taxonomy/compositional_skills/extraction/email/markdown/qna.yaml --endpoint-url https://781d2e7c-us-east.lb.appdomain.cloud/v1 --model-family mixtral --num-instructions 1 --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1 --tls-insecure --model /instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1
You are using an aliased command, this will be deprecated in a future release. Please consider using `ilab data generate` instead
INFO 2024-07-10 01:41:49,866 utils.py:145: _init_num_threads Note: detected 80 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" environment variable.
INFO 2024-07-10 01:41:49,866 utils.py:148: _init_num_threads Note: NumExpr detected 80 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 16.
INFO 2024-07-10 01:41:49,866 utils.py:161: _init_num_threads NumExpr defaulting to 16 threads.
INFO 2024-07-10 01:41:50,061 config.py:58: <module> PyTorch version 2.3.1 available.
Generating synthetic data using '/instructlab/models/mistralai/Mixtral-8x7B-Instruct-v0.1' model, taxonomy:'taxonomy/compositional_skills/extraction/email/markdown/qna.yaml' against https://781d2e7c-us-east.lb.appdomain.cloud/v1 server
INFO 2024-07-10 01:41:50,741 generate_data.py:266: generate_data Synthesizing new instructions. If you aren't satisfied with the generated instructions, interrupt training (Ctrl-C) and try adjusting your YAML files. Adding more examples may help.
INFO 2024-07-10 01:41:50,751 pipeline.py:48: generate Running block: gen_skill_grounded
INFO 2024-07-10 01:41:50,752 pipeline.py:49: generate Dataset({
    features: ['task_description', 'seed_context', 'seed_question', 'seed_response'],
    num_rows: 3
})
INFO 2024-07-10 01:42:00,856 generate_data.py:294: generate_data Generated 1 samples
INFO 2024-07-10 01:42:00,857 generate_data.py:312: generate_data Generation took 10.14s
[root@rhel-ai-sdg-training-client-1 instructlabbigdisk]# 
```

**Comment by markmc** [comment]
Related - #113 and #114

**Comment by khaledsulayman** [comment]
@relyt0925 could you add a brief description explaining the motivation behind these changes?

**Comment by relyt0925** [comment]
@khaledsulayman I closed this for the issue https://github.com/instructlab/instructlab/issues/1647 which I believe @tiran has ran with and also describes the case (which is mainly being able to validate TLS connections with a teacher model using a cert from a private CA)

**Pull request closed without merging**